src/model/CoilUserDefined.hpp

```diff
@@ -0,0 +1,166 @@
+*
+*  (4) Other than as required in clauses (1) and (2), distributions in any form of modifications or other derivative works
+*  may not use the "OpenStudio" trademark, "OS", "os", or any other confusingly similar designation without specific prior
+*  written permission from Alliance for Sustainable Energy, LLC.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) AND ANY CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+*  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+*  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER(S), ANY CONTRIBUTORS, THE UNITED STATES GOVERNMENT, OR THE UNITED
+*  STATES DEPARTMENT OF ENERGY, NOR ANY OF THEIR EMPLOYEES, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#ifndef MODEL_COILUSERDEFINED_HPP
+#define MODEL_COILUSERDEFINED_HPP
+
+#include "ModelAPI.hpp"
+#include "WaterToAirComponent.hpp"
+#include "EnergyManagementSystemProgramCallingManager.hpp"
+#include "EnergyManagementSystemProgram.hpp"
+#include "EnergyManagementSystemActuator.hpp"
+
+namespace openstudio {
+namespace model {
+
+  class EnergyManagementSystemProgramCallingManager;
+  class EnergyManagementSystemProgram;
+  class EnergyManagementSystemActuator;
+  class ThermalZone;
+
+  namespace detail {
+
+    class CoilUserDefined_Impl;
+
+  }  // namespace detail
+
+  /** CoilUserDefined is a WaterToAirComponent that wraps the OpenStudio IDD object 'OS:Coil:UserDefined'. */
+  class MODEL_API CoilUserDefined : public WaterToAirComponent
+  {
+   public:
+    /** @name Constructors and Destructors */
+    //@{
+
+    explicit CoilUserDefined(const Model& model);
+
+    virtual ~CoilUserDefined() = default;
+
+    //@}
+
+    static IddObjectType iddObjectType();
+
+    /** @name Getters */
+    //@{
+
+    boost::optional<EnergyManagementSystemProgramCallingManager> overallModelSimulationProgramCallingManager() const;
+
+    boost::optional<EnergyManagementSystemProgramCallingManager> modelSetupandSizingProgramCallingManager() const;
+
+    boost::optional<EnergyManagementSystemProgram> overallSimulationProgram() const;
+
+    boost::optional<EnergyManagementSystemProgram> initializationSimulationProgram() const;
+
+    boost::optional<EnergyManagementSystemActuator> airOutletTemperatureActuator() const;
+
+    boost::optional<EnergyManagementSystemActuator> airOutletHumidityRatioActuator() const;
```

Comment:
All these fields should be required-field if you are able to instantiate them in the Ctor.

Meaning the IDD should have \required-field and these getters should **not** return optional<T> but T.

Reply:
That also means no resetXXX methods.

Reply:
Maybe there is an explanation I'm missing for why you made them optional<T> on purpose though?

Reply:
agreed